x/cscalist/keeper/proposal.go

```diff
@@ -0,0 +1,24 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/rarimo/rarimo-core/x/cscalist/types"
+)
+
+func (k Keeper) EditCSCAListProposal(ctx sdk.Context, proposal *types.EditCSCAListProposal) error {
+	params := k.GetParams(ctx)
+	k.RemoveTree(ctx) // safe while no errors are expected, otherwise think about a backup
+
+	// TODO @violog Use a different tree implementation to allow passport-related ZKPs
+	// Ensure that the tree root always matches the one in related smart-contract
+	var (
+		tree = Treap{k}
+		lcg  = LCG{k}
+	)
+	for _, leaf := range proposal.Leaves {
```

Comment:
it will be useful to store also leaves data as it is and define key derivation function. then proposals can contain op type (remove/edit/insert) and leaves list (not a keys list) or it can be three (two) different proposals remove_leaves, add_leaves, edit_leaves (that logic can be implemented in add_leaves).

Reply:
can be discussed

Reply:
@violog

Reply:
IRL decision: 2 proposals: one rebuilds the whole tree, another allows to add/remove leaves